cmd/pod-scaler/admission.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/test-infra/prow/interrupts"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	buildv1 "github.com/openshift/api/build/v1"
+	buildclientset "github.com/openshift/client-go/build/clientset/versioned/typed/build/v1"
+	buildclientv1 "github.com/openshift/client-go/build/clientset/versioned/typed/build/v1"
+
+	"github.com/openshift/ci-tools/pkg/steps"
+)
+
+func admit(port int) {
+	logger := logrus.WithField("component", "admission")
+	restConfig, err := rest.InClusterConfig()
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to load in-cluster config.")
+	}
+	client, err := buildclientset.NewForConfig(restConfig)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to construct client.")
+	}
+	mgr, err := manager.New(restConfig, manager.Options{Port: port})
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not start manager.")
+	}
+	server := mgr.GetWebhookServer()
+	server.Register("/pods", &webhook.Admission{Handler: &podMutator{logger: logger, client: client}})
+	if err := mgr.Start(interrupts.Context()); err != nil {
+		logrus.WithError(err).Error("Failed to serve admission responses.")
+	}
+	logger.Debug("Ready to serve HTTP requests.")
+}
+
+type podMutator struct {
+	logger  *logrus.Entry
+	client  buildclientv1.BuildV1Interface
+	decoder *admission.Decoder
+}
+
+func (m *podMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
```

Comment:
No tests at all?

Reply:
I really couldn't figure out how to sensibly test this without having to create my own Scheme, which I really would rather not have to do

Reply:
Without the scheme for instance you end up hand-crafting AdmissionRequests with raw Object bodies, but also needing to craft the decoded objects, then you get weird diffs in the thing anyway ... what I would love is an e2e that the HTTP endpoint works TBH

Reply:
how about just testing a `handle(ctx context.Context, p *corev1.Pod, req admission.Request) admission.Response` ?